File: backend/src/main/java/com/collab/service/OTService.java
package com.collab.service;

import com.collab.model.websocket.Operation;
import org.springframework.stereotype.Service;

@Service
public class OTService {

    public String apply(String original, Operation operation) {
        String content = original == null ? "" : original;
        if (operation == null || operation.getType() == null) {
            return content;
        }

        int position = Math.max(0, Math.min(operation.getPosition() == null ? 0 : operation.getPosition(), content.length()));

        return switch (operation.getType()) {
            case INSERT -> content.substring(0, position) + defaultValue(operation.getContent()) + content.substring(position);
            case DELETE -> {
                int length = operation.getLength() == null ? defaultValue(operation.getContent()).length() : operation.getLength();
                int end = Math.min(content.length(), position + Math.max(length, 0));
                yield content.substring(0, position) + content.substring(end);
            }
            case FULL_SYNC -> defaultValue(operation.getContent());
        };
    }

    public Operation transform(Operation incoming, Operation applied) {
        if (incoming == null || incoming.getType() == null || applied == null || applied.getType() == null) {
            return incoming;
        }

        Operation transformed = Operation.builder()
                .type(incoming.getType())
                .position(incoming.getPosition())
                .length(incoming.getLength())
                .content(incoming.getContent())
                .revision(incoming.getRevision())
                .clientId(incoming.getClientId())
                .build();

        int incomingPosition = safePosition(transformed.getPosition());
        int appliedPosition = safePosition(applied.getPosition());
        int appliedLength = operationLength(applied);

        if (applied.getType() == Operation.OperationType.INSERT) {
            if (incomingPosition > appliedPosition || sameInsertPosition(incoming, applied)) {
                transformed.setPosition(incomingPosition + appliedLength);
            }
            return transformed;
        }

        if (applied.getType() == Operation.OperationType.DELETE) {
            if (incomingPosition >= appliedPosition + appliedLength) {
                transformed.setPosition(Math.max(appliedPosition, incomingPosition - appliedLength));
                return transformed;
            }

            if (incomingPosition >= appliedPosition) {
                transformed.setPosition(appliedPosition);
                if (transformed.getType() == Operation.OperationType.DELETE) {
                    int overlap = Math.min(appliedPosition + appliedLength, incomingPosition + operationLength(incoming)) - incomingPosition;
                    transformed.setLength(Math.max(0, operationLength(incoming) - Math.max(overlap, 0)));
                }
            }
        }

        return transformed;
    }

    private boolean sameInsertPosition(Operation incoming, Operation applied) {
        return incoming.getType() == Operation.OperationType.INSERT
                && applied.getType() == Operation.OperationType.INSERT
                && safePosition(incoming.getPosition()) == safePosition(applied.getPosition())
                && compareClientId(incoming.getClientId(), applied.getClientId()) > 0;
    }

    private int compareClientId(String left, String right) {
        return defaultValue(left).compareTo(defaultValue(right));
    }

    private int operationLength(Operation operation) {
        if (operation == null) {
            return 0;
        }
        if (operation.getLength() != null) {
            return Math.max(0, operation.getLength());
        }
        return defaultValue(operation.getContent()).length();
    }

    private int safePosition(Integer position) {
        return Math.max(0, position == null ? 0 : position);
    }

    private String defaultValue(String value) {
        return value == null ? "" : value;
    }
}

File: python-backend/app/services/ot_service.py
class OTOperation:
    """Represents a single OT operation."""

    def __init__(self, type_: str, position: int = 0, length: int = 0,
                 content: str = "", revision: int = 0, client_id: str = ""):
        self.type = type_  # INSERT, DELETE, FULL_SYNC
        self.position = position
        self.length = length
        self.content = content or ""
        self.revision = revision
        self.client_id = client_id or ""


class OTService:
    """Operational Transformation service."""

    @staticmethod
    def apply(original: str, operation: OTOperation) -> str:
        content = original or ""
        if not operation.type:
            return content

        position = max(0, min(operation.position, len(content)))

        if operation.type == "INSERT":
            return content[:position] + operation.content + content[position:]

        elif operation.type == "DELETE":
            op_len = operation.length if operation.length else len(operation.content)
            end = min(len(content), position + max(op_len, 0))
            return content[:position] + content[end:]

        elif operation.type == "FULL_SYNC":
            return operation.content or ""

        return content

    @staticmethod
    def transform(incoming: OTOperation, applied: OTOperation) -> OTOperation:
        if not incoming.type or not applied.type:
            return incoming

        transformed = OTOperation(
            type_=incoming.type,
            position=incoming.position,
            length=incoming.length,
            content=incoming.content,
            revision=incoming.revision,
            client_id=incoming.client_id,
        )

        in_pos = max(0, transformed.position)
        ap_pos = max(0, applied.position)
        ap_len = applied.length if applied.length else len(applied.content)

        if applied.type == "INSERT":
            if in_pos > ap_pos or (incoming.type == "INSERT" and applied.type == "INSERT"
                                    and in_pos == ap_pos and (incoming.client_id or "") > (applied.client_id or "")):
                transformed.position = in_pos + ap_len
            return transformed

        if applied.type == "DELETE":
            if in_pos >= ap_pos + ap_len:
                transformed.position = max(ap_pos, in_pos - ap_len)
                return transformed

            if in_pos >= ap_pos:
                transformed.position = ap_pos
                if transformed.type == "DELETE":
                    overlap = min(ap_pos + ap_len, in_pos + (transformed.length or len(transformed.content))) - in_pos
                    transformed.length = max(0, (transformed.length or len(transformed.content)) - max(overlap, 0))

        return transformed
